### ISSUE TITLE
How to select control group

### ISSUE THREAD

**Issue by sophia-wright-blue** [dafiti/causalimpact#47]
thanks for releasing and maintaining this amazing library. I have time series data for about 20 entities, and I'd like to select 5 of those as a control group to compare the results with the test group - how do I do this in this library?

(I think the `R` version released by Google allows you to pick the most useful groups from a given data set by placing a spike-and-slab prior distribution on the regression coefficients.)

I'd greatly appreciate any guidance. Thanks,

**Comment by WillianFuks** [comment]
Hello @sophia-wright-blue ,

The spike-slab strategy used in the Google's package is basically a way to implement a linear regression through the scope of Bayesian modeling.

As you can see in their [paper](https://arxiv.org/pdf/1506.00356.pdf), equation 2.9 is the Bernoulli selector for which regressors should remain for extracting the posterior of the counterfactual and equation 2.10 implements the prior for each "Beta" value, which essentially results in a linear (Bayesian) regression.

In this package, the covariates weights are also derived from a linear regression (as it also implements equations 2.1 and 2.2 from Google's paper, i.e, Bayesian Structural Time Series model), which means it also detect which series in the control are not helping much in explaining the observed data.

In order to confirm this, you can run this code:

```
import numpy as np
import pandas as pd
from statsmodels.tsa.arima_process import ArmaProcess
from causalimpact import CausalImpact


np.random.seed(12345)
ar = np.r_[1, 0.9]
ma = np.array([1])
arma_process = ArmaProcess(ar, ma)
X = 100 + arma_process.generate_sample(nsample=100)
y = 1.2 * X + np.random.normal(size=100)
y[70:] += 5

data = pd.DataFrame({'y': y, 'X': X}, columns=['y', 'X'])
pre_period = [0, 69]
post_period = [70, 99]

ci = CausalImpact(data, pre_period, post_period)
print(ci.summary())
```

Which gives:

```
Posterior Inference {Causal Impact}
                          Average            Cumulative
Actual                    125.2              3756.9
Prediction (s.d.)         120.4 (0.3)        3611.8 (8.2)
95% CI                    [119.9, 121.0]     [3596.4, 3628.7]

Absolute effect (s.d.)    4.8 (0.3)          145.1 (8.2)
95% CI                    [4.3, 5.3]         [128.1, 160.5]

Relative effect (s.d.)    4.0% (0.2%)        4.0% (0.2%)
95% CI                    [3.5%, 4.4%]       [3.5%, 4.4%]

Posterior tail-area probability p: 0.0
Posterior prob. of a causal effect: 100.00%

For more details run the command: print(impact.summary('report'))
```

Now we can add some random non useful simulated covariate as well, such as:

    X2 = np.linspace(0, 1, 100) * 100
    data = pd.DataFrame({'y': y, 'X': X, 'X2': X2}, columns=['y', 'X', 'X2'])
    ci = CausalImpact(data, pre_period, post_period)

Which gives:

```
Posterior Inference {Causal Impact}
                          Average            Cumulative
Actual                    125.2              3756.9
Prediction (s.d.)         120.8 (0.2)        3623.1 (7.4)
95% CI                    [120.3, 121.3]     [3608.9, 3637.8]

Absolute effect (s.d.)    4.5 (0.2)          133.8 (7.4)
95% CI                    [4.0, 4.9]         [119.1, 148.0]

Relative effect (s.d.)    3.7% (0.2%)        3.7% (0.2%)
95% CI                    [3.3%, 4.1%]       [3.3%, 4.1%]

Posterior tail-area probability p: 0.0
Posterior prob. of a causal effect: 100.00%

For more details run the command: print(impact.summary('report'))
```

As we can see from both results, not really changed; same conclusions and observed effect.

We can further investigate what the package itself is finding underneath the values by running the following command:

    ci.trained_model.summary()

Which gives:

```
<class 'statsmodels.iolib.summary.Summary'>
"""
                        Unobserved Components Results                         
==============================================================================
Dep. Variable:                      y   No. Observations:                   70
Model:                    local level   Log Likelihood                 -21.840
Date:                Wed, 18 Sep 2019   AIC                             51.680
Time:                        13:37:18   BIC                             60.616
Sample:                             0   HQIC                            55.225
                                 - 70                                         
Covariance Type:                  opg                                         
====================================================================================
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
sigma2.irregular     0.1034      0.021      4.966      0.000       0.063       0.144
sigma2.level      6.944e-05      0.001      0.097      0.923      -0.001       0.001
beta.X               0.9500      0.047     20.288      0.000       0.858       1.042
beta.X2              0.0684      0.068      1.011      0.312      -0.064       0.201
===================================================================================
Ljung-Box (Q):                       38.90   Jarque-Bera (JB):                 1.22
Prob(Q):                              0.52   Prob(JB):                         0.54
Heteroskedasticity (H):               1.09   Skew:                            -0.21
Prob(H) (two-sided):                  0.84   Kurtosis:                         2.51
===================================================================================

Warnings:
[1] Covariance matrix calculated using the outer product of gradients (complex-step).
"""
```

Notice the column `P>|z|`: it refers to a one tail statistical hypothesis testing of randomly observing same results if null hypothesis (no impact) is true; as the value is 31% we reject H1. What this does is basically the same as what equation 2.9 from Google's paper do: remove the impact from the posterior predictions of this specific weight.

So in summary, you probably don't need to worry much about finding best control set of variables, the package itself will already identify it for you.

Maybe if you have lots of predictors and performance is becoming an issue it might help to select some and see their impact by analyzing the `trained_model` summary and `ci.trained_model.pvalues` but other than that it shouldn't be a source of concern.

Not quite sure if this is the answer you were looking for, let us know if this helped and if you have any other questions :).

**Comment by sophia-wright-blue** [comment]
that was an amazing response - thanks so much! this helps a lot - I'm eagerly looking forward to understanding this library and the model better!

a quick follow-up:

have you had a chance to work with the R Market Matching library developed by Kim Larsen (https://github.com/klarsen1/MarketMatching)? Do you know of an equivalent Python library or workflow that I could use?

Thanks again for the detailed response!

**Comment by WillianFuks** [comment]
Hi @sophia-wright-blue ,

Glad to know the answer helped :)

Didn't know so far this library you shared; it seems an interesting wrap up to the Google's package.

Maybe doing something similar for the Python implementation would be a cool project. Thanks for sharing!

If you find any other issues just let us know.

Best,

Will
